core/src/main/java/io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java

```diff
@@ -25,38 +25,83 @@
 import io.questdb.cairo.sql.NoRandomAccessRecordCursor;
 import io.questdb.cairo.sql.Record;
 import io.questdb.cairo.sql.RecordMetadata;
+import io.questdb.cairo.sql.TableReferenceOutOfDateException;
 import io.questdb.cairo.vm.Vm;
-import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.SqlException;
 import io.questdb.griffin.engine.table.parquet.PartitionDecoder;
-import io.questdb.std.*;
-import io.questdb.std.str.*;
+import io.questdb.griffin.engine.table.parquet.RowGroupBuffers;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.BinarySequence;
+import io.questdb.std.Chars;
+import io.questdb.std.DirectBinarySequence;
+import io.questdb.std.DirectIntList;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.Long256;
+import io.questdb.std.Long256Impl;
+import io.questdb.std.LongList;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Misc;
+import io.questdb.std.Numbers;
+import io.questdb.std.Unsafe;
+import io.questdb.std.str.CharSink;
+import io.questdb.std.str.DirectString;
+import io.questdb.std.str.LPSZ;
+import io.questdb.std.str.Utf8Sequence;
+import io.questdb.std.str.Utf8SplitString;
 import org.jetbrains.annotations.Nullable;
 
 public class ReadParquetRecordCursor implements NoRandomAccessRecordCursor {
+    private static final Log LOG = LogFactory.getLog(ReadParquetRecordCursor.class);
     private final LongList auxPtrs = new LongList();
-    private final LongList columnChunkBufferPtrs = new LongList();
+    private final DirectIntList columns;
     private final LongList dataPtrs = new LongList();
     private final PartitionDecoder decoder;
+    private final FilesFacade ff;
     private final RecordMetadata metadata;
     private final ParquetRecord record;
+    private final RowGroupBuffers rowGroupBuffers;
+    private long addr = 0;
     private int currentRowInRowGroup;
-    private int rowGroup;
+    private long fd = -1;
+    private long fileSize = 0;
+    private int rowGroupIndex;
     private long rowGroupRowCount;
 
     public ReadParquetRecordCursor(FilesFacade ff, RecordMetadata metadata) {
-        this.metadata = metadata;
-        this.decoder = new PartitionDecoder(ff);
-        this.record = new ParquetRecord();
+        try {
+            this.ff = ff;
+            this.metadata = metadata;
+            this.decoder = new PartitionDecoder();
+            this.rowGroupBuffers = new RowGroupBuffers(MemoryTag.NATIVE_PARQUET_PARTITION_DECODER);
+            this.columns = new DirectIntList(32, MemoryTag.NATIVE_DEFAULT);
+            this.record = new ParquetRecord();
+        } catch (Throwable th) {
+            close();
+            throw th;
+        }
     }
 
+    @Override
     public void close() {
         Misc.free(decoder);
+        Misc.free(rowGroupBuffers);
+        Misc.free(columns);
+        if (fd != -1) {
+            ff.close(fd);
+            fd = -1;
+        }
+        if (addr != -1) {
```

Comment:
should this be `addr != 0` ?

Reply:
Good spot. Looks like this double close protection didn't work. Fixed in ae281710007817b247610e9f758148640f1a07a7